packages/@aws-cdk/aws-elasticsearch/lib/domain.ts

```diff
@@ -0,0 +1,1556 @@
+   */
+  public readonly masterUserPassword?: cdk.SecretValue;
+
+
+  private readonly domain: CfnDomain;
+
+  constructor(scope: cdk.Construct, id: string, props: DomainProps) {
+    super(scope, id, {
+      physicalName: props.domainName,
+    });
+
+    const defaultInstanceType = 'r5.large.elasticsearch';
+
+    const dedicatedMasterType =
+      props.capacity?.masterNodeInstanceType?.toLowerCase() ??
+      defaultInstanceType;
+    const dedicatedMasterCount = props.capacity?.masterNodes ?? 0;
+    const dedicatedMasterEnabled = dedicatedMasterCount > 0;
+
+    const instanceType =
+      props.capacity?.dataNodeInstanceType?.toLowerCase() ??
+      defaultInstanceType;
+    const instanceCount = props.capacity?.dataNodes ?? 1;
+
+    const availabilityZoneCount =
+      props.zoneAwareness?.availabilityZoneCount ?? 2;
+
+    if (![2, 3].includes(availabilityZoneCount)) {
+      throw new Error('Invalid zone awareness configuration; availabilityZoneCount must be 2 or 3');
+    }
+
+    const zoneAwarenessEnabled =
+      props.zoneAwareness?.enabled ??
+      props.zoneAwareness?.availabilityZoneCount != null;
+
+    // If VPC options are supplied ensure that the number of subnets matches the number AZ
+    if (props.vpcOptions != null && zoneAwarenessEnabled &&
+      new Set(props.vpcOptions?.subnets.map((subnet) => subnet.availabilityZone)).size < availabilityZoneCount) {
+      throw new Error('When providing vpc options you need to provide a subnet for each AZ you are using');
+    };
+
+    if ([dedicatedMasterType, instanceType].some(t => !t.endsWith('.elasticsearch'))) {
+      throw new Error('Master and data node instance types must end with ".elasticsearch".');
+    }
+
+    const elasticsearchVersion = props.version.version;
+    const elasticsearchVersionNum = parseVersion(props.version);
+
+    if (
+      elasticsearchVersionNum <= 7.7 &&
+      ![
+        1.5, 2.3, 5.1, 5.3, 5.5, 5.6, 6.0,
+        6.2, 6.3, 6.4, 6.5, 6.7, 6.8, 7.1, 7.4,
+        7.7,
+      ].includes(elasticsearchVersionNum)
+    ) {
+      throw new Error(`Unknown Elasticsearch version: ${elasticsearchVersion}`);
+    }
+
+    const unsignedBasicAuthEnabled = props.useUnsignedBasicAuth ?? false;
+
+    if (unsignedBasicAuthEnabled) {
+      if (props.enforceHttps == false) {
+        throw new Error('You cannot disable HTTPS and use unsigned basic auth');
+      }
+      if (props.nodeToNodeEncryption == false) {
+        throw new Error('You cannot disable node to node encryption and use unsigned basic auth');
+      }
+      if (props.encryptionAtRest?.enabled == false) {
+        throw new Error('You cannot disable encryption at rest and use unsigned basic auth');
+      }
+    }
+
+    const unsignedAccessPolicy = new iam.PolicyStatement({
+      effect: iam.Effect.ALLOW,
+      actions: ['es:ESHttp*'],
+      principals: [new iam.Anyone()],
+      resources: [cdk.Lazy.stringValue({ produce: () => `${this.domainArn}/*` })],
```

Comment:
This creates a circular dependency. You could build the ARN yourself by doing `this.stack.formatArn` - but that would only work if the user supplied a domain name in the properties. 

When we first spoke about this I hadn't considered that if the domain name is not supplied, we leave the name generation to CloudFormation (as we should) - but this means we have no way of knowing the name before the domain is actually created. 

To solve this, lets just use an `AwsCustomResource` to update the `AccessPolicies` post creation.

```typescript
const accessPoliciesDocument = props.accessPolicies ?? new iam.PolicyDocument()

if (unsignedBasicAuthEnabled) {
  accessPoliciesDocument.addStatements(
    new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ['es:ESHttp*'],
      principals: [new iam.Anyone()],
      resources: [`${this.domainArn}/*`],
    }),
  )
}

const accessPolicies = new cr.AwsCustomResource(this, 'AccessPolicies', {
  onUpdate: {
    action: 'updateElasticsearchDomainConfig',
    service: 'ES',
    parameters: {
      DomainName: this.domainName,
      AccessPolicies: JSON.stringify(policyDocument.toJSON()),
    },
    // this is needed to limit the response body, otherwise it exceeds the CFN 4k limit 
    outputPath: 'DomainConfig.ElasticsearchClusterConfig.AccessPolicies',
    physicalResourceId: cr.PhysicalResourceId.of(`${this.domainName}AccessPolicies`),
  },
  policy: cr.AwsCustomResourcePolicy.fromSdkCalls({ resources: [this.domainArn] }),
});

accessPolicies.node.addDependency(this.domain);

```

Reply:
I have added a custom resource for updating access policy and now use that to add the access policy after domain creation.

Reply:
The custom resource is a bit awkward. Isn't it possible to avoid the circular dependency by constructing the ARN using CFN pseudo params like this?: 
```typescript
"arn:aws:es:${AWS::Region}:${AWS::AccountId}:domain/" + this.domainName + "/*"
```

Reply:
W cannot use `this.domainName` because unless explicitly specified, it is assigned by CloudFormation at deploy time.